intermediate_source/dynamic_quantization_bert_tutorial.rst

```diff
@@ -0,0 +1,550 @@
+   a corpus of sentence pairs automatically extracted from online news
+   sources, with human annotations of whether the sentences in the pair
+   are semantically equivalent. As the classes are imbalanced (68%
+   positive, 32% negative), we follow the common practice and report
+   `F1 score <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html>`_.
+   MRPC is a common NLP task for language pair classification, as shown
+   below.
+
+.. image:: /_static/img/bert.png
+
+
+1. Setup
+--------
+
+1.1 Install PyTorch and HuggingFace Transformers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To start this tutorial, let’s first follow the installation instructions
+in PyTorch `here <https://github.com/pytorch/pytorch/#installation>`_ and HuggingFace Github Repo `here <https://github.com/huggingface/transformers#installation>`_.
+In addition, we also install `scikit-learn <https://github.com/scikit-learn/scikit-learn>`_ package, as we will reuse its
+built-in F1 score calculation helper function.
+
+.. code:: shell
+
+   pip install sklearn
+   pip install transformers
+
+
+Because we will be using the experimental parts of the PyTorch, it is
+recommended to install the latest version of torch and torchvision. You
+can find the most recent instructions on local installation `here
+<https://pytorch.org/get-started/locally/>`_. For example, to install on
+Mac:
+
+.. code:: shell
+
+   yes y | pip uninstall torch tochvision
+   yes y | pip install --pre torch -f https://download.pytorch.org/whl/nightly/cu101/torch_nightly.html
+
+
+
+
+1.2 Import the necessary modules
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this step we import the necessary Python modules for the tutorial.
+
+.. code:: python
+
+    from __future__ import absolute_import, division, print_function
+
+    import logging
+    import numpy as np
+    import os
+    import random
+    import sys
+    import time
+    import torch
+
+    from argparse import Namespace
+    from torch.utils.data import (DataLoader, RandomSampler, SequentialSampler,
+                                  TensorDataset)
+    from tqdm import tqdm
+    from transformers import (BertConfig, BertForSequenceClassification, BertTokenizer,)
+    from transformers import glue_compute_metrics as compute_metrics
+    from transformers import glue_output_modes as output_modes
+    from transformers import glue_processors as processors
+    from transformers import glue_convert_examples_to_features as convert_examples_to_features
+
+    # Setup logging
+    logger = logging.getLogger(__name__)
+    logging.basicConfig(format = '%(asctime)s - %(levelname)s - %(name)s -   %(message)s',
+                        datefmt = '%m/%d/%Y %H:%M:%S',
+                        level = logging.WARN)
+
+    logging.getLogger("transformers.modeling_utils").setLevel(
+       logging.WARN)  # Reduce logging
+
+    print(torch.__version__)
+
```

Comment:
The helper functions are built-in in transformers library. We mainly use
the following helper functions: one for converting the text examples
into the feature vectors; The other one for measuring the F1 score of
the predicted result.

The `glue_convert_examples_to_features <https://github.com/huggingface/transformers/blob/master/transformers/data/processors/glue.py>`_ function converts the texts into input features:

-  Tokenize the input sequences;
-  Insert [CLS] in the beginning;
-  Insert [SEP] between the first sentence and the second sentence, and
   in the end;
-  Generate token type ids to indicate whether a token belongs to the
   first sequence or the second sequence.

The `glue_compute_metrics` function computes the  `F1 score <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html>`_
can be interpreted as a weighted average of the precision and recall,
where an F1 score reaches its best value at 1 and worst score at 0. The
relative contribution of precision and recall to the F1 score are equal.

-  The equation for the F1 score is:
.. math:: F1 = 2 * (\text{precision} * \text{recall}) / (\text{precision} + \text{recall})

Reply:
@jianyuh : Does this look ok?